hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/LazyDefiningSpecificMemberPms.java

```diff
@@ -75,47 +75,48 @@ public void init(@Nonnull Context context) throws Exception {
         ExpressionEvalContext eec = ExpressionEvalContext.from(context);
         Expression<?> partitionKeyExpr = null;
         if (partitionKeyExprSupplier != null) {
-            Object obj = partitionKeyExprSupplier.get();
-            if (!(obj instanceof Expression)) {
-                throw new RuntimeException("");
-            }
-            partitionKeyExpr = (Expression<?>) obj;
+            partitionKeyExpr = partitionKeyExprSupplier.get();
         }
 
-        // test path
-        if (context instanceof TestProcessorMetaSupplierContext) {
-            TestProcessorMetaSupplierContext ctx = (TestProcessorMetaSupplierContext) context;
-            this.partitionId = ctx.getNodeEngine().getPartitionService().getPartitionId(
-                    partitionArgIndex != null
-                            ? eec.getArgument(partitionArgIndex)
-                            : Objects.requireNonNull(partitionKeyExpr).eval(null, eec));
-        } else {
-            assert context instanceof Contexts.MetaSupplierCtx;
-            Contexts.MetaSupplierCtx ctx = (Contexts.MetaSupplierCtx) context;
-            this.partitionId = ctx.nodeEngine().getPartitionService()
-                    .getPartitionId(
-                            partitionArgIndex != null
-                                    ? eec.getArgument(partitionArgIndex)
-                                    : Objects.requireNonNull(partitionKeyExpr).eval(null, eec));
-        }
+        this.partitionId = Util.getNodeEngine(context.hazelcastInstance()).getPartitionService().getPartitionId(
+                partitionArgIndex != null
+                        ? eec.getArgument(partitionArgIndex)
+                        : Objects.requireNonNull(partitionKeyExpr).eval(null, eec));
         partitionAssignment = context.partitionAssignment();
     }
 
+    @Nonnull
     @Override
     public Function<? super Address, ? extends ProcessorSupplier> get(@Nonnull List<Address> addresses) {
         Address address = null;
         for (Entry<Address, int[]> entry : partitionAssignment.entrySet()) {
-            for (int pId : entry.getValue()) {
-                if (pId == partitionId) {
-                    address = entry.getKey();
-                    break;
-                }
+            if (Arrays.binarySearch(entry.getValue(), partitionId) >= 0) {
+                address = entry.getKey();
+                break;
             }
         }
-        if (address == null && !addresses.contains(address)) {
-            throw new JetException("Cluster does not contain the required member: " + memberAddress);
+        final Address finalAddress = address;
+        // ExpectNothingProcessorSupplier may be eliminated by partition pruning, if used by SQL.
```

Comment:
did you mean "Processor pruning" instead of "partition pruning"?

Reply:
this PMS did not create `ExpectNothingProcessorSupplier` before this PR. New logic seems fine, was the previous one broken?

Reply:
Yes, it was broken :(